azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,190 @@
+                # Any methods will be no-ops, and query evaluation will return an empty allowlist
+                # so no extensions will be allowed.
+                log_policy("Policy enforcement is unsupported on this platform.")
+                return
+
+            # Regorus import should only be attempted after completing the above checks within
+            # the configurator, but the module itself needs to be accessible outside this class.
+            global regorus  # pylint: disable=global-statement
+            import azurelinuxagent.ga.policy.regorus as regorus
+            PolicyEngineConfigurator._policy_enabled = True
+
+        except (ImportError, NameError) as ex:
+            log_policy("Error: Failed to import Regorus module and initialize policy engine. {0}".format(ex), is_success=False)
+        except Exception as ex:
+            log_policy("Error: Failed to enable policy enforcement. '{0}'".format(ex), is_success=False)
+        finally:
+            PolicyEngineConfigurator._initialized = True
+
+    @staticmethod
+    def is_policy_configured():
+        """Return True if the policy enforcement feature is configured/enabled by customer."""
+        # TODO: Remove conf flag post private preview and add other checks here.
+        return conf.get_extension_policy_enabled()
+
+    @staticmethod
+    def _is_policy_supported():
+        """Return True if the platform/distro supports policy enforcement."""
+        arch = platform.machine().lower()
+        if arch not in POLICY_SUPPORTED_ARCHITECTURE:
+            return False
+        distro_info = get_distro()
+        distro_name = distro_info[0]
+        try:
+            distro_version = FlexibleVersion(distro_info[1])
+        except ValueError:
+            raise ValueError
+
+        # Check if the distro is in the support matrix and if the version is supported
+        if distro_name in POLICY_SUPPORTED_DISTROS:
+            min_version = POLICY_SUPPORTED_DISTROS[distro_name]
+            return distro_version >= min_version
+        else:
+            return False
+
+    @staticmethod
+    def get_instance():
+        if PolicyEngineConfigurator._instance is None:
+            PolicyEngineConfigurator._instance = PolicyEngineConfigurator()
+        return PolicyEngineConfigurator._instance
+
+    @staticmethod
+    def get_policy_enabled():
+        return PolicyEngineConfigurator.get_instance()._policy_enabled
+
+
+class PolicyEngine(object):
+    """
+    Implements policy engine API. Class will always be initialized, but if the Regorus import fails,
+    all methods will be no-ops.
+
+    If any errors are thrown in regorus.py, they will be caught and handled here. add_policy, add_data,
+    and set_input will be no-ops, eval_query will return an empty dict
+    """
+    def __init__(self):
+        self._policy_engine_enabled = False
+        self._engine = None
+        try:
+            if PolicyEngineConfigurator.get_instance().get_policy_enabled():
```

Comment:
This usage is weird, because get_policy_enabled() is an static method, so one does not need an instance to invoke it...

Why the call to get_instance()?

This is related to one of my comments above about all methods/properties of PolicyEngineConfigurator being static, so there is no need for a singleton instance

Reply:
Removed PolicyEngineConfigurator